### ISSUE TITLE
Performance of transitivity checks

### ISSUE THREAD

**Issue by fischmat** [anno4j/anno4j#138]
The new `ValidatedTransaction` provides schema validation at commit time.
Among others the transitivity property must be validated for every property of type `owl:TransitiveProperty`.
This leads to very high delays when committing and thus validating a transaction, because for each node reachable from any node directly accessed by the transaction object, every node reachable through a transitive property path must be checked.
To make things worse SPARQL does not support property path expressions for unbound variables, which results in a lot of subsequent queries.
Also a single LDPath expressions can't be used, because they are not expressive enough.

In order to improve the performance of the validation, I will try the following:

Instead of the normal `ObjectConnection` the transaction could use a `TrackedObjectConnection` which duplicates every modification made into a special context (graph).
For SPARQL UPDATE queries the same query is also executed again for the special context.
SPARQL DELETE queries can be transformed to an INSERT query to the special context.

At commit time the set of all nodes in the special context should be exactly the set of nodes modified by the transaction (directly or with setter-methods). 
Then for each modified node an iteration backward and one forward until the next unmodified node is necessary in order to check transitivity.

This procedure doubles the cost for every modification done with the transaction, but drastically reduces costs for checking transitivity in only partially modified regions of the graph.

**Comment by fischmat** [comment]
The time needed for validating a graph was reduced by two improvements to the transaction:

First special "working" contexts were introduced in order to be able to determine the set of affected resources at commit time. `ContextAwareConnection` (base class of `ObjectConnection`) features the definition of a context, which receives all inserted statements regardless of their actual context. Furthermore a archive context can be defined, receiving all triples that were removed from any other context.
Currently this approach comes with a **drawback**: Because statements from any context are inserted into the working graph, information about their actual context is lost. So after commit all statements from the working context are copied to the default graph of the repository. 
This means that the `ValidatedTransaction` is agnostic of context. A possible "solution" would be passing the transaction an context in which the statements are ultimately inserted. Thus the user would have to create a transaction for each context he wants to edit.

```java
// Possible solution. Creating a transaction for a specific target context:
ValidatedTransaction transaction = anno4j.createValidatedTransaction(myContext);
```

Second  a new approach was used to improve performance of the transitivity check.
Instead of recursively checking all edges using multiple queries, the check now uses a single query.
The query performs the transitivity check inductively and its runtime is linear dependent on the number of actually affected resources (not the graph size as before).

**Comment by fischmat** [comment]
It turned out that tracking removed statements using a archive context does not work out in Anno4j.
Thus the tracking of resources affected by a transaction is now implemented (see feb8292d8fa7d1754d876e4ad0ae9bae5e7ea4ff) using the set of cached resources of the `ObjectConnection` (which is probably even more elegant).
These cached resources and their direct neighbours are considered for validation, which enables short validation times.

Note:
Unfortunately the `ValidatedTransaction` inherits the ignoring of context from the normal `Transaction` (see #140).

**Comment by fischmat** [comment]
Short node on the implementation of the transitivity check for later reference:

The check is performed inductively and requires that the repository was consistent regarding the schema before the transaction started.
At commit time all affected resources and their direct neighborhood are identified (s.a.).
Now for each node `X` with a transitive property `p` the following is checked:
`∀ Y : (X, p, Y) ⇒ p(Y) ⊆ p(X)` with `p(X)` the set of objects of triples `(X, p, ?o)`. 
If the check is satisfied for `X` then for any `Y` holds:
- `Y` was not affected by the transaction and thus is valid by assumption.
- `Y` was affected by the transaction and
-- `Y` is valid then `X` is also valid, because `X` it has all further edges plus the one to `Y`
-- `Y` is invalid then the condition is satisfied for `X` but will fail for `Y` (induction)
